### PR TITLE
Flow cleanup, Google Test

### PR DESCRIPTION
**Pull request by shaomeng**
This PR cleans up some flow code, including fixing a pointer consistency issue.

### CODE REVIEW

**Comment by NihanthCW** [comment]
Remove Google Tests integration for now until further discussion.

**Comment by shaomeng** [comment]
> Remove Google Tests integration for now until further discussion.

This PR is ready to be reviewed without GTest.

File: lib/flow/VaporField.cpp [hunk 1]
```diff
@@ -575,4 +562,14 @@ const VAPoR::Grid *VaporField::_getAGrid(size_t timestep, const std::string &var
     return grid;
 }
 
-void VaporField::ReleaseLockedGrids() { _recentGrids.clear(); }
+
+void VaporField::ReleaseLockedGrids()
+{
+    // Release locked grids by giving the cache a bunch of nullptrs with unique invalid keys.
```
Comment (StasJ) [inline lib/flow/VaporField.cpp:568]: This seems like a hack.

**Review by StasJ — CHANGES_REQUESTED** [review]

**Review by StasJ — APPROVED** [review]

File: include/vapor/VaporField.h [hunk 2]
```diff
@@ -1,28 +1,29 @@
...
     std::string           _varName;
-    uint64_t              _timestep;
-    int32_t               _refLev, _compLev;
-    std::array<double, 3> _ext_min, _ext_max;
+    VAPoR::CoordType      _ext_min = {0.0, 0.0, 0.0};
+    VAPoR::CoordType      _ext_max = {0.0, 0.0, 0.0};
 
 public:
-    void Reset(uint64_t, int32_t, int32_t, std::string, const std::vector<double> &, const std::vector<double> &);
+    void Reset(uint32_t, int32_t, int32_t, std::string, VAPoR::CoordType, VAPoR::CoordType);
```
Comment (sgpearse) [inline include/vapor/VaporField.h:26]: Why switch from passing const references to non-const CoordType values?

File: include/vapor/ptr_cache.hpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// This is an implementation of a least-recently-used (LRU) cache that keeps
+// raw pointers pointing to big structures (e.g., grids, quadtrees).
+//
+// This cache has two execution policies:
+// 1) only insertion counts as `recently used`, and
+// 2) both insertion and query count as `recently used`.
+//
+// Given this design, this cache is expected to keep the ownership of these
+// structures once they're put in the cache, and all other codes will not
+// need to manage these structures. In other words, once an object is evected,
```
Comment (sgpearse) [inline include/vapor/ptr_cache.hpp:11]: Typo: Do you mean evicted instead of evected?

**Review by sgpearse — COMMENTED** [review]
Interesting stuff.  Just a question and I think I found a typo.

File: include/vapor/VaporField.h (hunk 2 above)
Comment (shaomeng) [reply include/vapor/VaporField.h:26]: In this case, I was attempting to make the function signature simpler (i.e., without using const and references).

Const references avoid a copy, which can be expensive if 1) the copy happens on the heap, and 2) the copy is big. In this case, after changing the signature, a copy happens on the stack and quite small with just 3 values, so it's not expensive anyways. I think the simplicity wins in this case.

File: include/vapor/ptr_cache.hpp (hunk 3 above)
Comment (shaomeng) [reply include/vapor/ptr_cache.hpp]: Good catch; should be fixed!

File: include/vapor/VaporField.h (hunk 2 above)
Comment (sgpearse) [reply include/vapor/VaporField.h:26]: Gotcha.  I was suprised to see this but if there's not a big performance hit then readability wins.

**Review by sgpearse — APPROVED** [review]

File: include/vapor/VaporField.h (hunk 2 above)
Comment (shaomeng) [reply include/vapor/VaporField.h:26]: Yeah, references have its own overheads, so for small amounts of data (debatable how small is small), simply make a copy seems easiest for everyone (including the compiler).

**Comment by shaomeng** [comment]
This PR is approved, I'll merge it to main so that Leigh can try it out!

**Pull request merged**